controller/volume_controller.go

```diff
@@ -808,12 +876,40 @@ func (vc *VolumeController) cleanupExtraHealthyReplicas(v *longhorn.Volume, e *l
 		for _, rName := range rNames {
 			r := rs[rName]
 			if r.Spec.NodeID != e.Spec.NodeID {
-				return vc.deleteReplica(r, rs)
+				if err := vc.deleteReplica(r, rs); err != nil {
+					return false, err
+				}
+				return true, nil
 			}
 		}
 	}
+	return false, nil
+}
 
-	return nil
+func (vc *VolumeController) getAutoBalancedReplicasSetting(v *longhorn.Volume) (types.ReplicaAutoBalance, error) {
+	var setting types.ReplicaAutoBalance
+
+	volumeSetting := v.Spec.ReplicaAutoBalance
+	if volumeSetting != types.ReplicaAutoBalanceIgnored {
+		setting = volumeSetting
+	}
+
+	var err error
+	if setting == "" {
+		globalSetting, _ := vc.ds.GetSettingValueExisted(types.SettingNameReplicaAutoBalance)
+
+		if globalSetting == string(types.ReplicaAutoBalanceIgnored) {
+			globalSetting = string(types.ReplicaAutoBalanceDisabled)
+		}
+
+		setting = types.ReplicaAutoBalance(globalSetting)
+	}
+
+	err = types.ValidateReplicaAutoBalance(types.ReplicaAutoBalance(setting))
+	if err != nil {
+		setting = types.ReplicaAutoBalanceDisabled
+	}
+	return setting, errors.Wrapf(err, "replica auto-balance is disabled")
```

Comment:
> If volumeSetting or globalSetting is not the defined option, should this function fall back to return types.ReplicaAutoBalanceDisabled (with a warning) rather than blindly convert the string to type types.ReplicaAutoBalance?

Addressed, thanks! (I don't know why I can't reply to the original comment)